### ISSUE TITLE
Rule proposal: No top-level side effects 

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1661]
### Description

_Opinionated, difficult/impossible to implement, and likely super annoying in practice._

Reasoning: importing a file should not have side effects

Related rules: 
- https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-unassigned-import.md
- https://eslint.org/docs/rules/no-new

### Fail

```js
document.title = 'gone'
```

### Pass

```js
const details = new Map();
```

```js 
function init () {
  document.title = 'gone'
}
```

```js
function init () {
  document.title = 'gone'
}
```

```js
function init () {
  document.title = 'gone'
}

// Still a side effect, but at least it's wrapped
// and not lost between many function declarations.
// Maybe a single, unassigned function call per file could be allowed
init(); 
```

```js
// Assignments could be considered side-effect free for the sake of usability
const details = new Map();
const response = fetch(); // even if they're not, really
```

**Comment by sindresorhus** [comment]
The rule should ignore files with a hash bang `#!/usr/bin/env node` and files with no exports. I think that will solve most of the potential annoyance.

**Comment by sindresorhus** [comment]
Rule name: `no-top-level-side-effects`?

**Comment by sindresorhus** [comment]
>  Maybe a single, unassigned function call per file could be allowed

**Comment by sindresorhus** [comment]
> Maybe a single, unassigned function call per file could be allowed

I don't think this exception is needed. You would usually only do this in CLIs or scripts, which are caught by https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1661#issuecomment-1055436984.

**Comment by sindresorhus** [comment]
> Assignments could be considered side-effect free for the sake of usability

Yes, assignments must be allowed.

**Comment by sindresorhus** [comment]
> Some exceptions could be made by filename

I don't think the rule should do that. You can use rule settings to ignore for certain files yourself.

**Comment by fregante** [comment]
> > Maybe a single, unassigned function call per file could be allowed
> 
> I don't think this exception is needed

How do I init my app? `ReactDOM.render(<App/>)` must exist somewhere and it cannot be `disable-next-line`’d every time.

**Comment by sindresorhus** [comment]
> How do I init my app? ReactDOM.render(<App/>) must exist somewhere and it cannot be disable-next-line’d every time, this is why at least one function call should be allowed. Not every JS file is an inert module.

I covered that:

> The rule should ignore files with a hash bang #!/usr/bin/env node **and files with no exports**.

**Comment by sindresorhus** [comment]
Accepted

**Comment by papb** [comment]
> Opinionated, difficult/impossible to implement, and likely super annoying in practice.

I agree with this. Any ideas on how to make this rule not have lots of annoying false positives?

**Comment by fregante** [comment]
Should this be called `pure-module`? Or `pure-top-level`? 

Related nomenclature: https://webpack.js.org/guides/tree-shaking/
